### PR TITLE
Correct cmakedefine implementation

### PR DESCRIPTION
**Pull request by Jan200101**
Changes the implementation behind `cmakedefine` and `cmakedefine01` to match the actual cmake behavior.

Tested against https://github.com/Jan200101/cmakedefine-test, which can configure a test header against cmake, meson and zig.

The diff between the cmake and zig configured header with this PR
```patch
--- cmake_build/test.h  2023-06-14 22:55:53.531659521 +0200
+++ zig-cache/o/9dcd70028e62d4109280ca26a82c002c/config.h       2023-06-14 22:53:41.225032597 +0200
@@ -1,3 +1,4 @@
+/* This file was generated by ConfigHeader using the Zig Build System. */

 // defined as "foo_content"
 #define foo "foo_content"
@@ -46,3 +47,4 @@

 // defined as 0, no value
 #define nofoo 0
+
```

Currently only supports `@var@` substition and not `${var}`.

`@var@` substition is also implemented in a very messy way, and I don't what would be the best way to clean it up, feedback on how to improve it would be useful.

### CODE REVIEW

**Comment by matu3ba** [comment]
This sounds like something, for which a standalone test would be nice to see what stuff is handled and what edge cases not.

**Comment by Jan200101** [comment]
variable substition with `@VAR@` and `${VAR}` is now implemented.

Its been separated into a function since #15017 could reuse this.

Tests will come next.

**Comment by Jan200101** [comment]
CI broke because of #16046, will fix whenever

**Comment by Jan200101** [comment]
standalone test added that compares an example header against a pre-configured header generated using cmake 3.26.4.

Overall the logic may be crusty on some ends, but this appears to work pretty well.

**Review by shubham-attri — APPROVED** [review]

**Comment by andrewrk** [comment]
Thanks! Good work
